src/ucp/core/ucp_ep.c

```diff
@@ -1137,20 +1155,38 @@ static void ucp_ep_set_lanes_failed(ucp_ep_h ep, uct_ep_h *uct_eps)
     }
 }
 
-static void ucp_ep_discard_lanes_callback(void *request, ucs_status_t status,
-                                          void *user_data)
+static void ucp_ep_discard_lanes(ucp_ep_h ep, ucs_status_t discard_status)
 {
-    ucp_ep_discard_lanes_arg_t *arg = (ucp_ep_discard_lanes_arg_t*)user_data;
+    unsigned ep_flush_flags         = (ucp_ep_config(ep)->key.err_mode ==
+                                       UCP_ERR_HANDLING_MODE_NONE) ?
+                                      UCT_FLUSH_FLAG_LOCAL :
+                                      UCT_FLUSH_FLAG_CANCEL;
+    uct_ep_h uct_eps[UCP_MAX_LANES] = { NULL };
+    ucs_status_t UCS_V_UNUSED status;
+    ucp_lane_index_t lane;
+    uct_ep_h uct_ep;
 
-    ucs_assert(arg != NULL);
-    ucs_assert(arg->counter > 0);
+    ucs_debug("ep %p: discarding lanes", ep);
 
-    if (--arg->counter > 0) {
-        return;
-    }
+    ucp_ep_set_lanes_failed(ep, uct_eps);
+
+    for (lane = 0; lane < ucp_ep_num_lanes(ep); ++lane) {
+        uct_ep = uct_eps[lane];
+        if (uct_ep == NULL) {
+            continue;
+        }
 
-    ucp_ep_reqs_purge(arg->ucp_ep, arg->status);
-    ucs_free(arg);
+        ucs_debug("ep %p: discard uct_ep[%d]=%p", ep, lane, uct_ep);
+
+        status = ucp_worker_discard_uct_ep(ep, uct_ep, ep_flush_flags,
+                                           ucp_ep_err_pending_purge,
+                                           UCS_STATUS_PTR(discard_status),
+                                           (ucp_send_nbx_callback_t)
+                                                   ucs_empty_function,
+                                           NULL);
```

Comment:
can we remove discard callback and arg from the parameters of ucp_worker_discard_uct_ep()? are they still needed?

Reply:
yes, this is still needed for wireup EP